sys/arm64/arm64/exception-morello.S

```diff
@@ -0,0 +1,259 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+#include <machine/asm.h>
+#include <machine/armreg.h>
+__FBSDID("$FreeBSD$");
+
+#include "assym.inc"
+
+	.text
+
+.macro	save_registers el
+.if \el == 1
+	mov	x18, sp
+	sub	sp, sp, #128
+.endif
+	stp	c29, c30, [sp], #-32
+	sub	sp, sp, #(TF_SIZE)
+	stp	c28, c29, [sp, #(TF_X + 28 * 16)]
+	stp	c26, c27, [sp, #(TF_X + 26 * 16)]
+	stp	c24, c25, [sp, #(TF_X + 24 * 16)]
+	stp	c22, c23, [sp, #(TF_X + 22 * 16)]
+	stp	c20, c21, [sp, #(TF_X + 20 * 16)]
+	stp	c18, c19, [sp, #(TF_X + 18 * 16)]
+	stp	c16, c17, [sp, #(TF_X + 16 * 16)]
+	stp	c14, c15, [sp, #(TF_X + 14 * 16)]
+	stp	c12, c13, [sp, #(TF_X + 12 * 16)]
+	stp	c10, c11, [sp, #(TF_X + 10 * 16)]
+	stp	c8,  c9,  [sp, #(TF_X + 8  * 16)]
+	stp	c6,  c7,  [sp, #(TF_X + 6  * 16)]
+	stp	c4,  c5,  [sp, #(TF_X + 4  * 16)]
+	stp	c2,  c3,  [sp, #(TF_X + 2  * 16)]
+	stp	c0,  c1,  [sp, #(TF_X + 0  * 16)]
+	mrs	c10, celr_el1
+	mrs	x11, spsr_el1
+	mrs	x12, esr_el1
+.if \el == 0
+	mrs	c18, csp_el0
+.endif
+	str	c10, [sp, #(TF_ELR)]
+	stp	w11, w12, [sp, #(TF_SPSR)]
+	stp	c18, clr, [sp, #(TF_SP)]
+	mrs	x18, tpidr_el1
+	add	x29, sp, #(TF_SIZE)
+.if \el == 0
+	/* Apply the SSBD (CVE-2018-3639) workaround if needed */
+	ldr	x1, [x18, #PC_SSBD]
+	cbz	x1, 1f
+	mov	w0, #1
+	blr	x1
+1:
+
+	ldr	x0, [x18, #(PC_CURTHREAD)]
+	bl	dbg_monitor_enter
+.endif
+	msr	daifclr, #8	/* Enable the debug exception */
+.endm
+
+.macro	restore_registers el
+.if \el == 1
+	/*
+	 * Disable interrupts and debug exceptions, x18 may change in the
+	 * interrupt exception handler.  For EL0 exceptions, do_ast already
+	 * did this.
+	 */
+	msr	daifset, #10
+.endif
+.if \el == 0
+	ldr	x0, [x18, #PC_CURTHREAD]
+	mov	x1, sp
+	bl	dbg_monitor_exit
+
+	/* Remove the SSBD (CVE-2018-3639) workaround if needed */
+	ldr	x1, [x18, #PC_SSBD]
+	cbz	x1, 1f
+	mov	w0, #0
+	blr	x1
+1:
+.endif
+	ldp	c18, clr, [sp, #(TF_SP)]
+	ldr	c10, [sp, #(TF_ELR)]
+	/* TODO: Remove when the kernel understands capabilities */
+	cvtp	c10, x10
+	ldr	w11, [sp, #(TF_SPSR)]
+.if \el == 0
+	msr	csp_el0, c18
+.endif
+	msr	spsr_el1, x11
+	msr	celr_el1, c10
+	ldp	c0,  c1,  [sp, #(TF_X + 0  * 16)]
+	ldp	c2,  c3,  [sp, #(TF_X + 2  * 16)]
+	ldp	c4,  c5,  [sp, #(TF_X + 4  * 16)]
+	ldp	c6,  c7,  [sp, #(TF_X + 6  * 16)]
+	ldp	c8,  c9,  [sp, #(TF_X + 8  * 16)]
+	ldp	c10, c11, [sp, #(TF_X + 10 * 16)]
+	ldp	c12, c13, [sp, #(TF_X + 12 * 16)]
+	ldp	c14, c15, [sp, #(TF_X + 14 * 16)]
+	ldp	c16, c17, [sp, #(TF_X + 16 * 16)]
+.if \el == 0
+	/*
+	 * We only restore the callee saved registers when returning to
+	 * userland as they may have been updated by a system call or signal.
+	 */
```

Comment:
This decision will come back to bite any attempts at kernel compartmentalisation...